### ISSUE TITLE
transfer hangs in a loop

### ISSUE THREAD

**Issue by fabianrost84** [wolpi/prim-ftpd#52]
I'd like to be able to backup all my files via ftp. However, I failed to transfer all my files from my Motorola Moto E on my computer. I used FileZilla on Kubuntu 15.10. I tried 2 different WiFis. FileZilla tries to transfer 2 specific files. The transfer fails with a timeout, FileZilla tries again, fails again and so on. I logged the transfer with FileZilla and with primitive ftpd. I upload the last 100 lines from each.
[filezilla.txt](https://github.com/wolpi/prim-ftpd/files/90521/filezilla.txt)
[primt-ftp.txt](https://github.com/wolpi/prim-ftpd/files/90533/primt-ftp.txt)

Any ideas what I could do to make this work?

**Comment by licaon-kter** [comment]
SFTP?

**Comment by fabianrost84** [comment]
The transfer also hangs like that if I use this FTP server: https://github.com/ppareit/swiftp 

So it might actually not be an issue with prim-ftpd.

**Comment by fabianrost84** [comment]
No I had the issue with FTP.

**Comment by licaon-kter** [comment]
But can you test with SFTP too ?

**Comment by fabianrost84** [comment]
Sure. I tested with SFTP. Transfer works without failures. It also transfers the file that did not work with FTP. However, I have issue #51 with SFTP.

**Comment by wolpi** [comment]
The filezilla log you posted contains e.g. this path:
`/storage/sdcard0/Android/data/com.fsck.k9/files/dc469241-93b6-45f5-be72-e08fabc16918.db_att/477`

The directory `/storage/sdcard0/Android/data/` was introduced sometime as android 4.x or 5 was the latest. It is declared as app-private. That means in the full path above only k9 is allowed to read/write, general file-manger (or (s)ftp server) apps are not allowed to access it.

These permissions are implemented differntly on each device. That you can change to those directories and see which file are present could be considered a bug in that specific android version.

Here is a link to google developer documentation. The section "Saving files that are app-private" is the one that affects this issue. But I fear it does not help you.
https://developer.android.com/guide/topics/data/data-storage.html


So how can you backup your files?
You cannot with this app.
You need root access to your device and you need an app that is able to act as root.

**Comment by licaon-kter** [comment]
`/storage/sdcard0/` are user accessible folders, they don't *need* root.

**Comment by wolpi** [comment]
That was true until maybe andorid 4.3 (not sure which version exactly). Now stuff that belonged to `/data/data/org.app.name` can be stored in `/storage/sdcard0/Android/data/org.app.name`. And to be precise: the base path `/storage/sdcard0` can differ on devices.

With this app you can access `/storage/sdcard0/` but you cannot `/storage/sdcard0/Android`. At least that is what google wants. Not all devices implement that and those that do do it differently. In this case here subdirectories and filenames are visible but files cannot be opened. On other devices the whole dir `/storage/sdcard0/Android` is hidden.

In the link you posted it is called "Application External storage directories".

**Comment by licaon-kter** [comment]
On 5.1.1 `just tested this and it's working correctly through primivitive FTPd in both modes.`

Anyway, @fabianrost84 using an explorer type app ( ES Explorer or TotalCommander or other) you can't copy a file from `/storage/sdcard0/Android/data/com.fsck.k9/files/dc469241-93b6-45f5-be72-e08fabc16918.db_att/477` to `/storage/sdcard0/Download` ?

**Comment by wolpi** [comment]
Should not work as those apps have same limited permissions as this one.
But please test to be sure.


> On 5.1.1 just tested this and it's working correctly through primivitive FTPd in both modes.

That means your device does not enforce those permissions.

**Comment by licaon-kter** [comment]
Tested so far:
* Xperia 5.1.1 (they tend to be close to AOSP) works ok.
* Samsung 5.0.2 works ok.
* CM 12.1 (might not enforce everything Google wants) works ok.

**Comment by fabianrost84** [comment]
In the end I used [SSHelper](http://arachnoid.com/android/SSHelper/index.html) to set up an SSH server on my phone and to backup files using rsync. I could transfer all files from `/storage/sdcard0/` using that app. Afterwards I performed a factory reset on my Moto E (this was my plan anyways and that's why I wanted to perform a backup). Afterwards, I tried again to backup`/storage/sdcard0/` using FileZilla. Now it works fine. There are again files under `/storage/sdcard0/Android/data/com.fsck.k9/files/`. I can transfer them all and also read the with an explorer type app on my phone. 

Thanks for all the support!
